src/coreComponents/finiteElement/CMakeLists.txt

```diff
@@ -37,12 +39,22 @@ endif()
 blt_add_library( NAME                  finiteElement
                  SOURCES               ${finiteElement_sources}
                  HEADERS               ${finiteElement_headers}
-                 DEPENDS_ON            ${dependencyList}
+                 DEFINES               JITTI_OUTPUT_DIR=${CMAKE_BINARY_DIR}/lib/jitti ${JITTI_DEFINES}
+                 DEPENDS_ON            ${dependencyList} jitti
                  OBJECT                ${GEOSX_BUILD_OBJ_LIBS}
                )
                
 target_include_directories( finiteElement PUBLIC ${CMAKE_SOURCE_DIR}/coreComponents)
 
+add_custom_command( OUTPUT ${CMAKE_BINARY_DIR}/include/kernelJITCompileCommands.hpp
+                    WORKING_DIRECTORY ${CMAKE_BINARY_DIR} 
+                    COMMAND python ${CMAKE_CURRENT_LIST_DIR}/../LvArray/src/jitti/generateCompileCommandsHeader.py 
+                                   ${CMAKE_BINARY_DIR}/compile_commands.json
+                                   --cpp ${CMAKE_CURRENT_LIST_DIR}/kernelInterface/kernelJIT.cpp 
+                                   --hpp ${CMAKE_BINARY_DIR}/include/kernelJITCompileCommands.hpp
+                                   --include ${CMAKE_BINARY_DIR}/include 
+                                   --linker ${CMAKE_CXX_COMPILER} )
```

Comment:
Instead of relying on headers that you generate on the file system, could it have been more robust to store the C++ pre-processing output as a string alongside the compile/link commands directly into the lib?
And that would be more "modern" JIT style imho.

Reply:
Yeah I think you could do that, although it would require the user to list out the functions that they want to JIT at configuration time in CMake. Then for each of those functions we could pre-process `jitti/templateSource.cpp`. But only up to a point, since we don't know what template params the user will want. Then at run time when we know the params we could compile the pre-processed source with the additional command line definitions of `JITTI_TEMPLATE_PARAMS` and `JITTI_TEMPLATE_PARAMS_STRING`.

However, how we pass a string embedded in our library to the compiler without using the file system is beyond me. Not to mention that at the end of the day the compiler is going to spit out a library on the file system that we then have to open. So this frees us from having to have access to the same headers used to build, but unless we do something really fancy I think file system access is a requirement.

Also at least with CUDA the cost of the compilation time itself will greatly outweigh the cost of opening and pre-processing the source.

Reply:
> Also at least with CUDA the cost of the compilation time itself will greatly outweigh the cost of opening and pre-processing the source.

It's more a problem of having a consistent self contained GEOSX installation than a performance problem imho.
Embedding the pre-processed source files directly into GEOSX would help at non relying on sources that may be modified behind the scenes.
I do not know all the JIT details so maybe it's not a problem anyway: I do not want to solve problems that do not exist.